### PR TITLE
Exceptions

### PR DESCRIPTION
**Pull request by tomphp**

### CODE REVIEW

File: src/Exception/NoMatchingFilesException.php [hunk 1]
```diff
@@ -2,6 +2,22 @@
 
 namespace TomPHP\ConfigServiceProvider\Exception;
 
-final class NoMatchingFilesException extends RuntimeException
+final class NoMatchingFilesException extends \LogicException implements Exception
```
Comment (localheinz) [inline src/Exception/NoMatchingFilesException.php]: @tomphp 

For consistency, should we import `LogicException` here?

File: tests/Exception/ReadOnlyExceptionTest.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+final class ReadOnlyExceptionTest extends PHPUnit_Framework_TestCase
+{
+    public function testItImplementsTheBaseExceptionType()
+    {
+        $this->assertInstanceOf(
+            'TomPHP\ConfigServiceProvider\Exception\Exception',
+            new ReadOnlyException()
+        );
+    }
```
Comment (localheinz) [inline tests/Exception/ReadOnlyExceptionTest.php]: @tomphp 

Not sure, maybe worth extracting an `AbstractExceptionTestCase`?

```php
class AbstractExceptionTestCase
{
    /**
     * @return string
     */
    abstract protected function className();

    final public function testItImplementsTheBaseExceptionType()
    {
        $exception = new $this->className();

        $this->assertInstanceOf('TomPHP\ConfigServiceProvider\Exception\Exception', $exception);
    }
}
```

But maybe overkill, hehe.

Comment (tomphp) [inline tests/Exception/ReadOnlyExceptionTest.php]: @localheinz I did actually consider it ;-) Might be worth doing at some point.

File: tests/Exception/UnknownFileTypeExceptionTest.php [hunk 3]
```diff
@@ -7,11 +7,27 @@
...
+
+    public function testItCanBeCreatedFromFileExtension()
+    {
+        $this->assertEquals(
+            'No reader configured for ".yml" files; readers are available for [".json", ".php"].',
+            UnknownFileTypeException::fromFileExtension(
+                '.yml',
+                ['.json', '.php']
+            )->getMessage()
+        );
```
Comment (localheinz) [inline tests/Exception/UnknownFileTypeExceptionTest.php]: @tomphp 

The wrapping seems a bit off, maybe worth extracting a variable?

```php
$exception = UnknownFileTypeException::fromFileExtension(
    '.yml',
    ['.json', '.php']
);

$this->assertEquals(
    'No reader configured for ".yml" files; readers are available for [".json", ".php"].', 
    $exception->getMessage()
);
```

?

Not sure what your stance is in regard to assertions, I usually try to be as strict as possible - that is, unless I really don't care about the returned type, I'd always assert sameness - what do you think?

**Comment by localheinz** [comment]
@tomphp 

👍

File: tests/Exception/UnknownFileTypeExceptionTest.php (hunk 3 above)
Comment (tomphp) [inline tests/Exception/UnknownFileTypeExceptionTest.php]: @localheinz agreed. I guess I've just got out of the habit as I tend to use PhpSpec usually and `shouldReturn` checks sameness.

**Comment by tomphp** [comment]
@localheinz merging - thanks for the review!

**Pull request merged**